url70ed135da7_ralfbiedert_jcores_tgz-pJ8-sandbox_SimpleFilterTestJ8/src/net/jcores/jre/cores/Core.java

```diff
@@ -0,0 +1,414 @@
+
+        // Level 2, dist = 4
+        // 0 x 4 -> 4
+        // Content: i b d d h i h h i i
+
+        // Return index[0]
+        
+        
+        final Runnable runner = new Runnable() {
+            public void run() {
+                int rnd = round.get();
+                int elementDistance = (int) Math.pow(2, rnd);
+
+                // Each thread processes as long as the distance between elements is smaller
+                // than the size
+                while (elementDistance < size) {
+                	// Each round the upper bound gets lower as we don't have to consider
+                	// the last elements
+                    final int upperBound = size - elementDistance;
+
+                    
+                    // Now we take a next element pair 
+                    int lastHandledI = -1;
+                    int i = baseCount.getAndAdd(2) * elementDistance;
+                    int j = i + elementDistance;
+
+                    // And each thread loop in here, until the right element
+                    // has left the righter bound
+                    while (j <= upperBound) {
+                    	// Process two elements and store them
+                        folder.handle(i, j, i);
+
+                        // Remember what we processed last
+                        lastHandledI = i;
+                        
+                        // Take the next available element pairs
+                        i = baseCount.getAndAdd(2) * elementDistance;
+                        j = i + elementDistance;
+                    }
+
+                    
+                    // Ideally, at this point, we are through the array. However, it might be the case 
+                    // that an uneven number of elements was in that loop run
+                    // 10 -> 5 pairs
+                    // 11 -> 5 pairs, one left
+                    
+                    // So we check now if we were the thread processing the last element pair
+                    // in that turn, and if there was a sole element left
+                    
+                    // This is the case when we processed something (lastHandledI > 0)
+                    // and when element 2nd element to the right (the next one) exists
+                    // but its pair (the 3rd one) does not. 
+                    
+                    // In that case we connect this element wit the next one 
+                    
+                    if (lastHandledI + 2 * elementDistance <= upperBound && 
+                        	lastHandledI + 3 * elementDistance > upperBound &&
+                        	lastHandledI >= 0) {
+                        	folder.handle(lastHandledI, lastHandledI + 2 * elementDistance, lastHandledI);
+                    }
+                
+                    // At this point we finished the round
+                    try {
+                        levelbarrier.await();
+                    } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)